sandbox/cgroup/monitor.go

```diff
@@ -0,0 +1,194 @@
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package cgroup
+
+import (
+	"errors"
+	"path"
+	"sync"
+
+	"github.com/snapcore/snapd/osutil"
+	"github.com/snapcore/snapd/osutil/inotify"
+)
+
+// Manages the inotify watcher, allowing to have a single watch descriptor open
+type inotifyWatcher struct {
+	// The watch object
+	wd *inotify.Watcher
+	// This is used to ensure that the watcher goroutine is launched only once
+	doOnce sync.Once
+	// This channel is used to add a new CGroup that has to be checked
+	addWatchChan chan *groupToWatch
+	// Contains the list of groups to monitor, to detect when they have been deleted
+	groupList []*groupToWatch
+	// Contains the list of paths being monitored by the inotify watcher.
+	// The paths monitored aren't the CGroup paths, but the parent ones, because
+	// in /sys/fs is not possible to detect the InDeleteSelf message, only the
+	// InDelete one, so we must monitor the parent folder and detect when a
+	// specific folder, corresponding to a CGroup, has been deleted.
+	// Each path has associated an integer, which is an usage counter. Every
+	// time a new CGroup paths have to be monitored, the parent paths are
+	// checked in the map, and if they already exist, the usage counter is
+	// incremented by one (of course, if a parent path isn't in the map,
+	// it is initialized to 'one'). Every time the path of a CGroup is deleted,
+	// the corresponding usage counter here corresponding to the parent path
+	// is decremented, and when it reaches zero, it is removed and inotify is
+	// notified to stop monitoring it.
+	pathList map[string]int32
+}
+
+// Contains the data corresponding to a CGroup that must be watched to detect
+// when it is destroyed
+type groupToWatch struct {
+	// This is the CGroup identifier
+	name string
+	// This contains a list of folders to monitor. When all of them have been
+	// deleted, the CGroup has been destroyed and there are no processes running
+	folders []string
+	// This channel is used to notify to the requester that this CGroup has been
+	// destroyed. The watcher writes the CGroup identifier on it; this way, the
+	// same channel can be shared to monitor several CGroups.
+	channel chan string
+}
+
+var currentWatcher *inotifyWatcher = &inotifyWatcher{
+	wd:           nil,
+	pathList:     make(map[string]int32),
+	addWatchChan: make(chan *groupToWatch),
+}
+
+func (iw *inotifyWatcher) addWatch(newWatch *groupToWatch) {
+	var folderList []string
+	for _, fullPath := range newWatch.folders {
+		// It's not possible to use inotify.InDeleteSelf in /sys/fs because it
+		// isn't triggered, so we must monitor the parent folder and use InDelete
+		basePath := path.Dir(fullPath)
+		if _, exists := iw.pathList[basePath]; !exists {
+			iw.pathList[basePath] = 0
+			if err := iw.wd.AddWatch(basePath, inotify.InDelete); err != nil {
```

Comment:
Yes, because the error happens when adding a non-existing path, which happens when the CGroup was removed between the call to "monitor" and this. That's why at line 97 I check if the length is zero and send the "completed" notification in that case.